### PR TITLE
Feature blocks

### PR DESCRIPTION
**Pull request by guusw**
Implements chainblocks => shader translation.

Add Feature blocks with the following:
- Render state
- Shader entry points
- Shader parameters

Linked to #171 & #168

### CODE REVIEW

File: include/chainblocks.hpp
Comment (sinkingsugar) [inline include/chainblocks.hpp:953]: Didn't we have this? can't rem

File: src/core/blocks/casting.hpp
Comment (sinkingsugar) [inline src/core/blocks/casting.hpp:1]: Use header guard

File: src/core/blocks/math.hpp
Comment (sinkingsugar) [inline src/core/blocks/math.hpp:89]: this was long overdue.. but not sure why I never did it... for some reason prevented something. If it works cool though :)

File: include/chainblocks.hpp
Comment (guusw) [reply include/chainblocks.hpp:953]: Didn't seem to be there for sequences

File: src/core/blocks/math.hpp
Comment (guusw) [reply src/core/blocks/math.hpp:89]: I noticed the broadcast operators were missing when trying to translate them for shaders. decided to add them 😄

File: src/core/blocks/casting.hpp
Comment (guusw) [reply src/core/blocks/casting.hpp]: done

**Comment by sinkingsugar** [comment]
Ready to review?

File: src/extra/gfx/shader/translator.cpp
Comment (sinkingsugar) [inline src/extra/gfx/shader/translator.cpp:183]: Do we need `Shader.` prefix I wonder 🤔

Comment (guusw) [reply src/extra/gfx/shader/translator.cpp:183]: I wanted to avoid it using the context mechanism (#167) where they would only be available inside shader wires.
But for now I think it's better to avoid the name collisions like this.

File: src/tests/gfx_feature.edn
Comment (Kryptos-FR) [inline src/tests/gfx_feature.edn:8]: You can use `(defshards)` here as an alias (and avoid the additional `(->)`.

```clj
(defshards spin-transform [t location]
```

File: src/tests/gfx_feature.edn
Comment (Kryptos-FR) [inline src/tests/gfx_feature.edn:12]: Temporary variable can be skipped using [`(Sub)`](https://docs.fragcolor.xyz/shards/General/Sub/) (or its alias `(|)`.

```suggestion
   t
   (| (Math.Multiply 0.2) (Math.AxisAngleX) (Math.Rotation) >= .rotX)
   (| (Math.Multiply 0.7) (Math.AxisAngleY) (Math.Rotation) >= .rotY)
   (| (Math.Multiply 0.9) (Math.AxisAngleZ) (Math.Rotation) >= .rotZ)
```

File: src/core/shards/casting.cpp
Comment (Kryptos-FR) [inline src/core/shards/casting.cpp:324]: Is that a failed merge conflict resolution? I believe all mention of "cb" should have been replaced by "sh".

File: src/core/shards/casting.hpp
Comment (Kryptos-FR) [inline src/core/shards/casting.hpp:198]: Missing end of line

**Review by Kryptos-FR — CHANGES_REQUESTED** [review]

**Review by sinkingsugar — APPROVED** [review]

File: src/core/shards/casting.cpp
Comment (sinkingsugar) [reply src/core/shards/casting.cpp:324]: yep 👍

**Comment by sinkingsugar** [comment]
Again less codecov could be a red flag

File: src/tests/gfx_feature.edn
Comment (guusw) [reply src/tests/gfx_feature.edn:12]: Forgot about that one, thanks!

**Comment by guusw** [comment]
Waiting for test run with GPU tests, will merge after.

**Review by Kryptos-FR — APPROVED** [review]

**Pull request merged**
